RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducer.cc

```diff
@@ -8,55 +8,80 @@
       int nRH = 0;
       if (pfRecHits->metadata().size() != 0)
         nRH = pfRecHits->size();
 
-      reco::PFClusteringVarsDeviceCollection pfClusteringVars{nRH, event.queue()};
-      reco::PFClusteringEdgeVarsDeviceCollection pfClusteringEdgeVars{(nRH * 8), event.queue()};
-      reco::PFClusterDeviceCollection pfClusters{nRH, event.queue()};
-      reco::PFRecHitFractionDeviceCollection pfrhFractions{nRH * pfRecHitFractionAllocation_, event.queue()};
+      pfClusteringVars_ = std::make_optional<reco::PFClusteringVarsDeviceCollection>(nRH, event.queue());
+      pfClusteringEdgeVars_ = std::make_optional<reco::PFClusteringEdgeVarsDeviceCollection>(nRH * 8, event.queue());
+      pfClusters_ = std::make_optional<reco::PFClusterDeviceCollection>(nRH, event.queue());
+
+      *num_rhf_ = 0;
+
+      if (nRH != 0) {
+        PFClusterProducerKernel kernel(event.queue(), pfRecHits);
+        kernel.step1(event.queue(),
+                     params,
+                     topology,
+                     *pfClusteringVars_,
+                     *pfClusteringEdgeVars_,
+                     pfRecHits,
+                     *pfClusters_,
+                     num_rhf_.data());
+      }
+    }
+
+    void produce(device::Event& event, device::EventSetup const& setup) override {
+      const reco::PFClusterParamsDeviceCollection& params = setup.getData(pfClusParamsToken);
+      const reco::PFRecHitHCALTopologyDeviceCollection& topology = setup.getData(topologyToken_);
+      const reco::PFRecHitHostCollection& pfRecHits = event.get(inputPFRecHitSoA_Token_);
+      int nRH = 0;
+      if (pfRecHits->metadata().size() != 0)
+        nRH = pfRecHits->size();
 
       if (nRH != 0) {
+        pfrhFractions_ = std::make_optional<reco::PFRecHitFractionDeviceCollection>(*num_rhf_.data(), event.queue());
         PFClusterProducerKernel kernel(event.queue(), pfRecHits);
-        kernel.execute(event.queue(),
-                       params,
-                       topology,
-                       pfClusteringVars,
-                       pfClusteringEdgeVars,
-                       pfRecHits,
-                       pfClusters,
-                       pfrhFractions);
+        kernel.step2(event.queue(),
+                     params,
+                     topology,
+                     *pfClusteringVars_,
+                     *pfClusteringEdgeVars_,
+                     pfRecHits,
+                     *pfClusters_,
+                     *pfrhFractions_);
+      } else {
+        pfrhFractions_ = std::make_optional<reco::PFRecHitFractionDeviceCollection>(0, event.queue());
       }
 
       if (synchronise_)
         alpaka::wait(event.queue());
 
-      event.emplace(outputPFClusterSoA_Token_, std::move(pfClusters));
-      event.emplace(outputPFRHFractionSoA_Token_, std::move(pfrhFractions));
+      event.emplace(outputPFClusterSoA_Token_, std::move(*pfClusters_));
+      event.emplace(outputPFRHFractionSoA_Token_, std::move(*pfrhFractions_));
```

Comment:
@makortel would it be better to `reset()` the optionals at the end of `produce()` ?

Reply:
> Although probably there is no meaningful difference between a "full" `std::optional` that containes a moved-from collection, or an "empty" `std::optional`.

There is a difference that in the latter case `std::optional::has_value()` returns `false`, whereas in the former case it returns `true`. But in the present code, that doesn't use `has_value()` or `bool` conversion operator, there indeed is no practical difference between the two (but if someone later makes use of the `has_value()`, then the difference would become relevant).